trie-db/src/lookup.rs

```diff
@@ -655,4 +668,175 @@ where
+			loop {
+				let decoded = match L::Codec::decode(node_data) {
+					Ok(node) => node,
+					Err(e) => return Err(Box::new(TrieError::DecoderError(hash, e))),
+				};
+
+				let next_node = match decoded {
+					Node::Leaf(slice, _) => {
+						// The leaf slice can be longer than remainder of the provided key
+						// (descendent), but not the other way around.
+						if !slice.starts_with(&partial) {
+							self.record(|| TrieAccess::NonExisting { full_key });
+							return Ok(None)
+						}
+
+						if partial.len() != slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+						}
+
+						return Ok(Some(hash))
+					},
+					Node::Extension(slice, item) => {
+						let common_prefix_len = partial.common_prefix(&slice);
+
+						let start_matches = if partial.len() <= slice.len() {
+							// Extension slice can be longer than remainder of the provided key
+							// (descendent), ensure the extension slice starts with the remainder
+							// of the provided key.
+							//
+							// This effectively returns the hash of the `Node::Branch` at the next
+							// iteration.
+							common_prefix_len == partial.len()
+						} else {
+							// Remainder of the provided key is longer than the extension slice,
+							// must advance the node iteration if and only if keys share
+							// a common prefix.
+							common_prefix_len == slice.len()
+						};
+
+						if start_matches {
+							// Empties the partial key if the extension slice is longer.
+							partial = partial.mid(common_prefix_len);
+							key_nibbles += common_prefix_len;
+							item
+						} else {
+							self.record(|| TrieAccess::NonExisting { full_key });
+
+							return Ok(None)
+						}
+					},
+					Node::Branch(children, value) =>
+						if partial.is_empty() {
+							if value.is_none() {
+								self.record(|| TrieAccess::NonExisting { full_key });
+							}
+
+							return Ok(Some(hash))
+						} else {
+							match children[partial.at(0) as usize] {
+								Some(x) => {
+									partial = partial.mid(1);
+									key_nibbles += 1;
+									x
+								},
+								None => {
+									self.record(|| TrieAccess::NonExisting { full_key });
+
+									return Ok(None)
+								},
+							}
+						},
+					Node::NibbledBranch(slice, children, value) => {
+						let common_prefix_len = partial.common_prefix(&slice);
+						// Not enough remainder key to continue the search.
+						if partial.len() < slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+
+							// Branch slice starts with the remainder key, there's nothing to
+							// advance.
+							if common_prefix_len == partial.len() {
+								return Ok(Some(hash))
+							} else {
+								return Ok(None)
+							}
+						}
+
+						// Partial key is longer or equal than the branch slice.
+						// Ensure partial key starts with the branch slice.
+						if common_prefix_len != slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
```

Comment:
what if common_prefix_len is less than slice.len() and at the same time common_prefix_len is equal to partial.len().
Aka the queried key is into the partial, then I think we need to return the hash of the current branch.
(similar case as starts_with in leaf).

Reply:
Would this now be the case of `partial.len() < slice.len() { if slice.starts_with(&partial) } `?

Have changed the code to use `starts_with` instead of the common length prefix, which makes it easier to read for me, thanks!